### PR TITLE
Adding a Channel example

### PR DESCRIPTION
**Pull request by lukecheeseman**
Building a channel from cowns, promises and queues.

### CODE REVIEW

File: testsuite/demo/run-pass/channel.verona [hunk 1]
```diff
@@ -0,0 +1,156 @@
+/*****************************************************************************
+ * This example demonstrates how one could build a channel.
+ *
```
Comment (sophiaIC) [inline testsuite/demo/run-pass/channel.verona]: This example demonstrates how one could build channels in Verona using cowns

File: testsuite/demo/run-pass/channel.verona [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * It illustrates creating and using promises and cowns and communicating
+ * between behaviours.
```
Comment (sophiaIC) [inline testsuite/demo/run-pass/channel.verona]: Does it?

**Comment by mjp41** [comment]
CI requires you put at least one
```C++
   // CHECK: ...
```
or
```C++
   //CHECK-L: ... 
```
Preferably, add something that actually checks it does what you expect, if possible.

File: testsuite/demo/run-pass/channel.verona (hunk 2 above)
Comment (sophiaIC) [reply testsuite/demo/run-pass/channel.verona]: As well as cowns, the implementation uses promises to communicate between behaviours.

File: testsuite/demo/run-pass/channel.verona [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+ *****************************************************************************/
+
+use "library/queue.verona"
+
+/*
+ * Create a channel from two queues:
+ *   - A queue of reads
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: Are both queues required at the same time? Or could you play the same trick as in `Scheduler`?  It is fine to be different, just want to understand.

File: testsuite/demo/run-pass/channel.verona [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  create(data: Cell[T] & iso): Writer[T] & iso {
+    var result = new Writer;
+    result._data = data;
+    result
+  }
+
+  read(self: iso): (Cell[T] & iso) | (None & imm) {
+    self._data = Builtin.freeze(new None)
+  }
+}
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: I would like these classes to appear above Channel.  So I can read it from top to bottom and understand the types.

File: testsuite/demo/run-pass/channel.verona [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  /*
+   * Write is a synchronous method.
+   * The method schedules a behaviour to log a write
+   * to the channel.
+   */
+  write(channel: cown[Channel[T]] & imm, value: T) {
+    var cell = Cell.create(value);
+    when(channel) {
+      match((mut-view (channel._rs)).remove()) {
+        var _: None => (mut-view (channel._ws)).add(Writer.create(cell)),
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: Add comments to explain the two cases. 
```
// No readers are waiting, so add this `value` to the end of the queue
```
and
```
// Fulfill the longest waiting reader.
```

File: testsuite/demo/run-pass/channel.verona [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+   */
+  read(channel: cown[Channel[T]] & imm): cown[Cell[T]] & imm {
+    var r = Reader.create();
+    var handle = (mut-view r).await();
+    when(channel) {
+      match((mut-view (channel._ws)).remove()) {
+        var _: None => (mut-view (channel._rs)).add(r),
+        var w: Writer[T] => {
+          match (w.read()) {
+            var _: None => {} /* shouldn't be possible */
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: Perhaps elaborate on why?

File: testsuite/demo/run-pass/channel.verona [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+   * Read is an asynchronous method.
+   * The method returns the wait part of a promise
+   * that will be fulfilled when a value is available.
+   */
+  read(channel: cown[Channel[T]] & imm): cown[Cell[T]] & imm {
+    var r = Reader.create();
+    var handle = (mut-view r).await();
+    when(channel) {
+      match((mut-view (channel._ws)).remove()) {
+        var _: None => (mut-view (channel._rs)).add(r),
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: Explain two cases here.

File: testsuite/demo/run-pass/channel.verona [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /*
+   * Read is an asynchronous method.
+   * The method returns the wait part of a promise
+   * that will be fulfilled when a value is available.
+   */
+  read(channel: cown[Channel[T]] & imm): cown[Cell[T]] & imm {
+    var r = Reader.create();
+    var handle = (mut-view r).await();
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: This is nice as the code can return before this has even entered the promise in the queue, explain this.

File: testsuite/demo/run-pass/channel.verona [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+    when(var received = Channel.read(channel)) {
+      match (received.extract()) {
+        var _: None => Builtin.print("Empty\n"),
+        var o: U64Obj => Builtin.print1("Contents: {:#}\n", o),
+      }
+    };
+
+    Channel.write(channel, U64Obj.create(52));
+  }
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: Add 
// CHECK-L: Contents: 7
// CHECK-L: Contents 52

File: testsuite/demo/run-pass/channel.verona (hunk 3 above)
Comment (lukecheeseman) [reply testsuite/demo/run-pass/channel.verona]: You can play the same trick, I experimented with the single queue. This is, in some ways, just to be different but also the implementation with a single queue makes the read/write methods larger. The extra logic for changing the queue type distracts a little from the meaning of the read/write methods.

File: testsuite/demo/run-pass/channel.verona [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    var channel = Channel.create();
+
+    /* write a message into the channel */
+    Channel.write(channel, U64Obj.create(7));
+
+    /* wait for a message to be received on the channel */
+    when(var received = Channel.read(channel)) {
+      match (received.extract()) {
+        var _: None => Builtin.print("Empty\n"),
+        var o: U64Obj => Builtin.print1("Contents: {:#}\n", o.v),
```
Comment (mjp41) [inline testsuite/demo/run-pass/channel.verona]: The DAG aspect is interesting.  Could you change this to 
"Contents 1:"
and the subsequent one to "Contents 2:"

I believe you would then have 
```
// CHECK-DAG: Contents 1: 7
// CHECK-DAG: Contents 2: 52
```
But wanted to clarify.  Also, refining the test will help with understanding.

Comment (lukecheeseman) [reply testsuite/demo/run-pass/channel.verona]: Yes, that would be the expected output. Thanks, those changes are a good idea.

Although, looking at the CI again, it uses OutputCheck and not FileCheck so I can't use CHECK-DAG.

Comment (mjp41) [reply testsuite/demo/run-pass/channel.verona]: You could just check one of them?

**Pull request merged**